File: royalroad.py
import random

fitness_calculator = None


class FitnessCalculator:
    def __init__(self, solution):
        self.solution = solution

    def score_fitness(self, individual):
    	"""
    	Compares the genes of a given individual to check them against the target individual, seeing how many match
    	"""
        return len([i for i, j in zip(self.solution, individual.genes) if i == j])

    def get_max_fitness(self):
    	"""
    	Curently hard coded to match Individuals
    	"""
        return 64


class Individual:
    def __init__(self, init):
        self.genes = []
        self.gene_length = 64
        self.generate_individual()

    def generate_individual(self):
        """
        Creates a randomized individual by filling its gene list with random 1s and 0s
        """
        for i in xrange(self.gene_length):
            self.genes.append(random.randint(0, 1))

    def get_gene(self, index):
        return self.genes[index]

    def set_gene(self, index, to_set):
        self.genes[index] = to_set;

    def size(self):
    	return len(self.genes)


class Population:
    def __init__(self, population_size, init):
        self.individuals = []

        if not init:
            return

        for i in xrange(population_size):
            self.individuals.append(Individual(True))

    def add_individual(self, individual):
        self.individuals.append(individual)

    def get_fittest(self):
    	"""
    	Finds the fittest individual in this Population by scoring them all
    	"""
        fittest = self.individuals[0]
        for individual in self.individuals:
            if fitness_calculator.score_fitness(individual) > fitness_calculator.score_fitness(fittest):
                fittest = individual
        return fittest

    def get_random(self):
    	"""
    	Gets a random member of this Population
    	"""
        return random.choice(self.individuals)

    def size(self):
        return len(self.individuals)


class Algorithm:
    def __init__(self):
        self.mutation_rate = 0.015
        self.uniformRate = 0.5
        self.elitism = True
        self.tournament_size = 5

    def evolve_population(self, population):
    	"""
    	Evolves a given population into a new one

    	It does this by conditionally preserving the fittest individual, and then
    	crossing over existing Individuals into new ones.  Finally, it may mutate
    	Individuals based on the Algorithm's mutation_rate.
    	"""
        new_population = Population(population.size(), False)
        elitism_offset = 0

        if self.elitism:
            new_population.add_individual(population.get_fittest())
            elitism_offset = 1

        for i in xrange(elitism_offset, population.size()):
            individual1 = self.tournament_selection(population)
            individual2 = self.tournament_selection(population)
            new_individual = self.crossover(individual1, individual2)
            new_population.add_individual(new_individual)

        for count, individual in enumerate(new_population.individuals):
            if self.elitism and count == 0:
                continue
            self.mutate(individual)

        return new_population

    def mutate(self, individual):
    	"""
    	Checks to see if a given Individual should be mutated
    	"""
        for i in xrange(individual.size()):
        	if random.random() <= self.mutation_rate:
        		individual.set_gene(i, random.randint(0, 1))
                
    def crossover(self, individual1, individual2):
    	"""
    	Crosses two Individuals into one new one
    	"""
        hybrid = Individual(False)

        for i in xrange(individual1.gene_length):
            if random.random() <= self.uniformRate:
                hybrid.set_gene(i, individual1.get_gene(i))
            else:
                hybrid.set_gene(i, individual2.get_gene(i))

        return hybrid

    def tournament_selection(self, population):
    	"""
    	Creates a temporary 'tournament' Population from which to select a strong
    	individual from.  The tournament Population is generally a subset of the
    	provided Population.
    	"""
        tournament = Population(self.tournament_size, False)
        for i in xrange(self.tournament_size):
            tournament.add_individual(population.get_random())

        return tournament.get_fittest()


if __name__ == '__main__':
    algorithm = Algorithm()
    target_individual = Individual(True)
    fitness_calculator = FitnessCalculator(target_individual.genes)

    population = Population(20, True)

    generation_count = 0

    while fitness_calculator.score_fitness(population.get_fittest()) != fitness_calculator.get_max_fitness():
        generation_count += 1
        population = algorithm.evolve_population(population)

    print 'answer found in generation: ' + str(generation_count)
    print population.get_fittest().genes

File: royalroad.go
package main

import "fmt"
import "math/rand"
import "time"

// Fitness
type FitnessCalculator struct {
	solution Individual
}

func (fc *FitnessCalculator) scoreFitness(ind Individual) int {
	score := 0
	for i := 0; i < 64; i++ {
		if fc.solution.genes[i] == ind.genes[i] {
			score++
		}
	}
	return score
}

func (fc *FitnessCalculator) getMaxFitness() int {
	return 64
}

// Individuals
type Individual struct {
	genes [64]int
}

func (ind *Individual) generateIndividual() {

	var buffer [64]int
	for i := 0; i < 64; i++ {
		buffer[i] = rand.Intn(2)
	}
	ind.genes = buffer
}

func (ind *Individual) getGene(index int) int {
	return ind.genes[index]
}

func (ind *Individual) setGene(index int, value int) {
	ind.genes[index] = value
}

func (ind *Individual) size() int {
	return len(ind.genes)
}

// Populations
type Population struct {
	individuals []Individual
}

func (pop *Population) initialize(size int, init bool) {
	if !init {
		return
	}
	for i := 0; i < size; i++ {
		ind := Individual{}
		ind.generateIndividual()
		pop.addIndividual(ind)
	}
}

func (pop *Population) addIndividual(ind Individual) {
	pop.individuals = append(pop.individuals, ind)
}

func (pop *Population) getFittest(fc FitnessCalculator) Individual {
	fittest := pop.individuals[0]
	for i := 0; i < len(pop.individuals); i++ {
		if fc.scoreFitness(pop.individuals[i]) > fc.scoreFitness(fittest) {
			fittest = pop.individuals[i]
		}
	}
	return fittest
}

func (pop *Population) getRandom() Individual {
	return pop.individuals[rand.Intn(20)]
}

func (pop *Population) size() int {
	return len(pop.individuals)
}

// Algorithms
type Algorithm struct {
	mutationRate   float32
	uniformRate    float32
	elitism        bool
	tournamentSize int
	fitnessCalc    FitnessCalculator
}

func (algor *Algorithm) evolvePopulation(pop Population) Population {
	newPopulation := Population{}
	elitismOffset := 0

	if algor.elitism {
		newPopulation.addIndividual(pop.getFittest(algor.fitnessCalc))
		elitismOffset = 1
	}

	for i := elitismOffset; i < pop.size(); i++ {
		individual1 := algor.tournamentSelection(pop)
		individual2 := algor.tournamentSelection(pop)
		newIndividual := algor.crossover(individual1, individual2)
		newPopulation.addIndividual(newIndividual)
	}

	for i := 0; i < newPopulation.size(); i++ {
		if algor.elitism && i == 0 {
			continue
		}
		algor.mutate(&newPopulation.individuals[i])
	}

	return newPopulation
}

func (algor *Algorithm) mutate(ind *Individual) {
	for i := 0; i < ind.size(); i++ {
		if rand.Float32() <= algor.mutationRate {
			ind.setGene(i, rand.Intn(2))
		}
	}
}

func (algor *Algorithm) crossover(individual1 Individual, individual2 Individual) Individual {
	hybrid := Individual{}
	for i := 0; i < individual1.size(); i++ {
		if rand.Float32() <= algor.uniformRate {
			hybrid.setGene(i, individual1.getGene(i))
		} else {
			hybrid.setGene(i, individual2.getGene(i))
		}
	}
	return hybrid
}

func (algor *Algorithm) tournamentSelection(pop Population) Individual {
	tournament := new(Population)
	for i := 0; i < algor.tournamentSize; i++ {
		tournament.addIndividual(pop.getRandom())
	}

	return tournament.getFittest(algor.fitnessCalc)
}

func main() {
	fmt.Println("Royal Road!")
	rand.Seed(time.Now().UTC().UnixNano())

	solution := Individual{}
	solution.generateIndividual()

	fitnessCalc := FitnessCalculator{solution}
	algorithm := Algorithm{0.015, 0.5, true, 5, fitnessCalc}

	generationCount := 0
	population := Population{}
	population.initialize(20, true)

	for {
		if fitnessCalc.scoreFitness(population.getFittest(fitnessCalc)) == fitnessCalc.getMaxFitness() {
			break
		}
		generationCount++
		population = algorithm.evolvePopulation(population)
		fmt.Println("generation & fitness", generationCount, fitnessCalc.scoreFitness(population.getFittest(fitnessCalc)))
	}
	fmt.Println("answer found in generation ", generationCount)
}
